### PR TITLE
[Storage] [DataMovement] Blob Preserve - Http headers, Access Tier and Metadata

### PR DESCRIPTION
**Pull request by amnguye**
Proposal how to expose preserving metadata and other properties on the storage resource. These API changes will also be done in File Shares.

### CODE REVIEW

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Metadata = System.Collections.Generic.IDictionary<string, string>;
+
+namespace Azure.Storage.DataMovement
+{
+    /// <summary>
+    ///  Represents the options for transferring metadata from the source to the destination storage resource.
+    /// </summary>
+    public class BlobTransferMetadataOptions
```
Comment (jaschrep-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs:14]: Can we not have both? A sort of "preserve metadata but add/overwrite these values" situation? Not sure we need to have exclusively one or the other.

**Comment by jaschrep-msft** [comment]
Something that's striking me reading these new options is that there isn't much nuance for the customer, given the "preserve" mode is something the library is handling for you, rather than a service feature. The one service that has preserve features built in does them on a per-field basis, which I like.

For metadata and other arbitrary-key properties, I wonder if we need to come up with a dictionary organization where we can say
1. no entry preserves
2. value entry creates/overwrites
3. null value deletes

For typed property collections like http headers, we may need a generic struct to use per-value. Something like `Preservable<T>` (some workshopping to do there) such that customers can declare explicitly if they're preserving or deleting, as we can't tell the difference from a null value otherwise.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs (hunk 1 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs:14]: You're right, I didn't think about that scenario. That someone wants to keep the existing metadata, and they want to add more metadata on.

**Comment by amnguye** [comment]
> Something that's striking me reading these new options is that there isn't much nuance for the customer, given the "preserve" mode is something the library is handling for you, rather than a service feature. The one service that has preserve features built in does them on a per-field basis, which I like.
> 
> For metadata and other arbitrary-key properties, I wonder if we need to come up with a dictionary organization where we can say
> 
> 1. no entry preserves
> 2. value entry creates/overwrites
> 3. null value deletes
> 
> For typed property collections like http headers, we may need a generic struct to use per-value. Something like `Preservable<T>` (some workshopping to do there) such that customers can declare explicitly if they're preserving or deleting, as we can't tell the difference from a null value otherwise.

I'm interested in this `Preservable<T>` kind of concept, just because creating a class for each and every "property" type gets very repetitive.

But I also want to make sure that this will be flexible in the future as well. Like in the future if we get a feature request to handle Invalid Metadata (like in AzCopy) to rename the metadata if invalid, fail or skip over it.

File: sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs [hunk 2]
```diff
@@ -2184,7 +2184,7 @@ public async Task StartCopyFromUriAsync_Source_IfTags()
         public async Task StartCopyFromUriAsync_Source_IfTagsFailed()
         {
             // Arrange
-            await using DisposingContainer test = await GetTestContainerAsync();
+            await using DisposingContainer test = await GetTestContainerAsync(publicAccessType: PublicAccessType.BlobContainer);
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs:2187]: TODO: undo this change

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferHttpHeadersOptions.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// Licensed under the MIT License.
+
+using Azure.Storage.Blobs.Models;
+
+namespace Azure.Storage.DataMovement
+{
+    /// <summary>
+    ///  Represents the options for transferring HttpHeaders from the source to the destination storage resource.
+    /// </summary>
+    public class BlobTransferHttpHeadersOptions
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferHttpHeadersOptions.cs:11]: Reminants of the old API implementation I made, TODO: remove

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+// Licensed under the MIT License.
+
+using Metadata = System.Collections.Generic.IDictionary<string, string>;
+
+namespace Azure.Storage.DataMovement
+{
+    /// <summary>
+    ///  Represents the options for transferring metadata from the source to the destination storage resource.
+    /// </summary>
+    public class BlobTransferMetadataOptions
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferMetadataOptions.cs:11]: Reminants of the old API implementation I made, TODO: remove

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferTagsOptions.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Tags = System.Collections.Generic.IDictionary<string, string>;
+
+namespace Azure.Storage.DataMovement
+{
+    /// <summary>
+    ///  Represents the options for transferring tags from the source to the destination storage resource.
+    /// </summary>
+    public class BlobTransferTagsOptions
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobTransferTagsOptions.cs:14]: Reminants of the old API implementation I made, TODO: remove

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs [hunk 6]
```diff
@@ -624,5 +628,148 @@ internal static StorageResourceItemProperties ToResourceProperties(this BlobItem
...
+            }
+            if (callSetHttpHeaders)
+            {
+                await destinationBlobClient.SetHttpHeadersAsync(headers).ConfigureAwait(false);
+            }
+
+            // If the Access Tier does NOT want to be preserved,
+            // and does not contain a value. Clear the access tier.
+            if (!(options?.AccessTier?.Preserve ?? true) &&
+                options?.AccessTier?.Value == default)
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs:710]: We probably can't "clear" access tier.

Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs]: Yup we can't. I removed this

File: sdk/storage/Azure.Storage.DataMovement/api/Azure.Storage.DataMovement.net6.0.cs [hunk 7]
```diff
@@ -156,7 +175,7 @@ public abstract partial class StorageResourceItem : Azure.Storage.DataMovement.S
         protected internal abstract string ResourceId { get; }
         protected Azure.Storage.DataMovement.StorageResourceItemProperties ResourceProperties { get { throw null; } set { } }
         protected internal abstract Azure.Storage.DataMovement.DataTransferOrder TransferType { get; }
-        protected internal abstract System.Threading.Tasks.Task CompleteTransferAsync(bool overwrite, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken));
+        protected internal abstract System.Threading.Tasks.Task CompleteTransferAsync(bool overwrite, Azure.Storage.DataMovement.StorageResourceItemProperties sourceProperties = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken));
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/api/Azure.Storage.DataMovement.net6.0.cs]: Done

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlockBlobStorageResource.cs [hunk 8]
```diff
@@ -204,8 +208,20 @@ internal class BlockBlobStorageResource : StorageResourceItemInternal
...
+                    _options,
+                    overwrite,
+                    options?.SourceAuthentication,
+                    options?.SourceProperties),
                 cancellationToken: cancellationToken).ConfigureAwait(false);
+            // By default, a Copy From URL call will automatically
+            // copy over the source properties to the destination blob.
+            // If the metdata does NOT want to be preserved,
+            // set them to the value specified.
+            await DataMovementBlobsExtensions.ClearMetadataIfSet(
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlockBlobStorageResource.cs]: Removed this. Figured out we could use the Copy Source Properties method, and then set some properties if we wanted them to be set.

This is done above in the ToBlobUploadFromUriOptions method.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs [hunk 9]
```diff
@@ -54,10 +54,13 @@ public partial class BlobStorageResourceContainerOptions
...
-        public Azure.Storage.Blobs.Models.AccessTier? AccessTier { get { throw null; } set { } }
-        public Azure.Storage.Blobs.Models.BlobHttpHeaders HttpHeaders { get { throw null; } set { } }
-        public System.Collections.Generic.IDictionary<string, string> Metadata { get { throw null; } set { } }
-        public System.Collections.Generic.IDictionary<string, string> Tags { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<Azure.Storage.Blobs.Models.AccessTier?> AccessTier { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<string> CacheControl { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<string> ContentDisposition { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<string> ContentEncoding { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<string> ContentLanguage { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<string> ContentType { get { throw null; } set { } }
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:62]: We talked about this offline. I think we prefer to leave these broken out, which will allow us more flexibility in the feature to allow certain options to be set, without making major API changes.

[the remark this answers is not in the document]
Comment (jalauzon-msft) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:62]: Two follow-up here:
- What happens if someone chooses to preserve some and not preserve others during single-call copy? Since there is only the one setting for all properties. This also applies to metadata. Combining all HTTP headers into one could at least make this less likely to happen but it could still happen with metadata and headers.
- I think we discussed adding `ContentHash` back in? I think we should do that if these stay as seperate settings.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs [hunk 10]
```diff
@@ -2,8 +2,13 @@
 // Licensed under the MIT License.
 
 using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
 using System.Collections.Generic;
 using System.IO;
+using System.Reflection.Emit;
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs:8]: nit: This seems like something that is not used?

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobDestinationCheckpointData.cs [hunk 11]
```diff
@@ -16,51 +16,90 @@ internal class BlobDestinationCheckpointData : BlobCheckpointData
...
-            Tags blobTags)
+            DataTransferProperty<string> contentType,
+            DataTransferProperty<string> contentEncoding,
+            DataTransferProperty<string> contentLanguage,
+            DataTransferProperty<string> contentDisposition,
+            DataTransferProperty<string> cacheControl,
+            DataTransferProperty<AccessTier?> accessTier,
+            DataTransferProperty<Metadata> metadata,
+            DataTransferProperty<Tags> tags)
             : base(DataMovementBlobConstants.DestinationCheckpointData.SchemaVersion, blobType)
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobDestinationCheckpointData.cs:71]: Since we released version 1 now, we probably need to bump the schema version for this to 2. Additionally, we'll have to put in checks to fail if we see version 1 now (since we're ok failing for old preview versions).

File: sdk/storage/Azure.Storage.DataMovement/src/ServiceToServiceJobPart.cs [hunk 12]
```diff
@@ -483,7 +503,11 @@ internal void DisposeHandlers()
 
         private async Task<StorageResourceCopyFromUriOptions> GetCopyFromUriOptionsAsync(CancellationToken cancellationToken)
         {
-            StorageResourceCopyFromUriOptions options = default;
+            StorageResourceItemProperties properties = await _sourceResource.GetPropertiesAsync(cancellationToken).ConfigureAwait(false);
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/ServiceToServiceJobPart.cs:506]: This will always be cached in the resource from initial call when initializing the job part right? If fo, maybe add a comment so we remember that in the future.

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs [hunk 13]
```diff
@@ -102,10 +102,14 @@ internal static void WritePaddedString(this BinaryWriter writer, string value, i
         /// <summary>
         /// Writes a boolean plus int32 to represent a nulable int.
         /// </summary>
-        internal static void Write(this BinaryWriter writer, int? value)
+        internal static void Write(
+            this BinaryWriter writer,
+            int? value,
+            ref int currentOffset)
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs:108]: All the other write extensions do not increment any sort of offset. Why did you add it to this one specifically? If it needs to stay, we should at least update the docstring to indicate it will increment the offset.

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs [hunk 14]
```diff
@@ -148,6 +152,16 @@ internal static void Write(this BinaryWriter writer, DateTimeOffset? value)
...
         }
 
+        /// <summary>
+        /// Writes two int zero values to represent the length and offset of a value
+        /// that is preserved.
+        /// </summary>
+        internal static void WriteEmptyLengthOffset(this BinaryWriter writer)
+        {
+            writer.Write(0);
+            writer.Write(0);
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs:162]: I think we should write the offset as `-1` to indicate it's not present. I thought I had this pattern already somewhere? Maybe in the job part schema or something.

File: sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs [hunk 15]
```diff
@@ -463,7 +464,11 @@ public async Task PauseThenResumeTransferAsync(TransferDirection transferType)
             await testEventsRaised.AssertPausedCheck();
 
             // Act - Resume Job
-            DataTransferOptions resumeOptions = new DataTransferOptions();
+            DataTransferOptions resumeOptions = new DataTransferOptions()
+            {
+                // Enable overwrite on resume, to overwrite destination.
+                CreationPreference = StorageResourceCreationPreference.OverwriteIfExists
+            };
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs:471]: Curious why you needed to add this? I tried to keep this out of these tests so we could test that we could get a "clean" pause/resume.

File: sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs [hunk 16]
```diff
@@ -842,8 +843,8 @@ public async Task TryPauseTransferAsync_DataTransfer_Directory(TransferDirection
             using DisposingLocalDirectory checkpointerDirectory = DisposingLocalDirectory.GetTestDirectory();
             using DisposingLocalDirectory sourceDirectory = DisposingLocalDirectory.GetTestDirectory();
             using DisposingLocalDirectory destinationDirectory = DisposingLocalDirectory.GetTestDirectory();
-            await using DisposingContainer sourceContainer = await GetTestContainerAsync();
-            await using DisposingContainer destinationContainer = await GetTestContainerAsync();
+            await using DisposingContainer sourceContainer = await GetTestContainerAsync(publicAccessType: PublicAccessType.BlobContainer);
+            await using DisposingContainer destinationContainer = await GetTestContainerAsync(publicAccessType: PublicAccessType.BlobContainer);
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs:847]: Usually destination containers should not need to be public? Same below.

File: sdk/storage/Azure.Storage.DataMovement/tests/Shared/StartTransferUploadTestBase.cs [hunk 17]
```diff
@@ -754,6 +754,6 @@ public async Task LocalToRemoteObject_LargeMultiple(int objectCount, long size,
                 waitTimeInSec: waitTimeInSec,
                 objectCount: objectCount);
         }
-#endregion
+        #endregion
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement/tests/Shared/StartTransferUploadTestBase.cs:757]: nit: I think this is not supposed to be indented?

File: sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:62]: - If someone decides to do that. Then we have to set the copy-properties header to false, and then set the properties that they did want to preserve manually. Basically we would be doing the same thing we would be doing on a chunked copy (aka preserving the properties by manually setting each property value). Same for metadata.
- If we combine it into one, it removes the flexibility in the future when we do get the ask to preserve some of the properties but not all. Which is the flexibility they have today in the base storage SDK.
- I will add `ContentHash` back in, but for future reference `ContentHash` is one of those properties that easily evolve into a feature ask to take many forms given our past interactions with customers regarding this property.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs [hunk 18]
```diff
@@ -54,10 +54,13 @@ public partial class BlobStorageResourceContainerOptions
     public partial class BlobStorageResourceOptions
     {
         public BlobStorageResourceOptions() { }
-        public Azure.Storage.Blobs.Models.AccessTier? AccessTier { get { throw null; } set { } }
-        public Azure.Storage.Blobs.Models.BlobHttpHeaders HttpHeaders { get { throw null; } set { } }
-        public System.Collections.Generic.IDictionary<string, string> Metadata { get { throw null; } set { } }
-        public System.Collections.Generic.IDictionary<string, string> Tags { get { throw null; } set { } }
+        public Azure.Storage.DataMovement.DataTransferProperty<Azure.Storage.Blobs.Models.AccessTier?> AccessTier { get { throw null; } set { } }
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:57]: I think you mean, we can't clear AccessTier right? I think we can keep this property and document a limitation since the service doesn't allow us to clear AccessTier.

File: sdk/storage/Azure.Storage.DataMovement/src/ServiceToServiceJobPart.cs (hunk 12 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/ServiceToServiceJobPart.cs:506]: Yes it will be. This is done on line 186 of this same file. I checked where this method was called previously, which should cache the properties on the source resource.

File: sdk/storage/Azure.Storage.DataMovement/tests/Shared/StartTransferUploadTestBase.cs (hunk 17 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/tests/Shared/StartTransferUploadTestBase.cs:757]: It was actually supposed to be indented. Cause the region it's encompassing is all indented.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:62]: I did the above with the exception of `ContentHash`. This will be done later on which we talked about offline.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobDestinationCheckpointData.cs (hunk 11 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobDestinationCheckpointData.cs:71]: Upped to 2 and also add tests for this.

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs (hunk 14 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs]: done

File: sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs (hunk 16 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs:847]: Removed for the source, left for the destination.

File: sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs (hunk 15 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/tests/PauseResumeTransferTests.cs:471]: We talked about this offline but looks like there's a bug with getting a clean pause and updated checkpointer file. I added a bug for that to address in a separate PR.

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs (hunk 13 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement/src/Shared/CheckpointerExtensions.cs:108]: Oh I think I added this as something to fix something else, and figured out I didn't need this. I'm going to remove this method all together since this is not being called anywhere.

Thanks for the call out.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs [hunk 19]
```diff
@@ -382,6 +378,23 @@ internal static BlockBlobStageBlockOptions ToBlobStageBlockOptions(this BlockBlo
                 DestinationConditions = CreateRequestConditions(options?.DestinationConditions, overwrite),
                 SourceAuthentication = sourceAuthorization,
             };
+            if ((options?.ContentEncoding?.Preserve ?? true) &&
+                (options?.ContentDisposition?.Preserve ?? true) &&
+                (options?.ContentLanguage?.Preserve ?? true) &&
+                (options?.ContentType?.Preserve ?? true) &&
+                (options?.CacheControl?.Preserve ?? true) &&
+                (options?.AccessTier?.Preserve ?? true) &&
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs:386]: Does Put Blob from URL always copy access tier? Or does it depend on `CopySourceBlobProperties`?

If it always copies it no matter what, I don't think we should have `AccessTier` here?

**Review by jalauzon-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs (hunk 19 above)
Comment (amnguye) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/DataMovementBlobsExtensions.cs:386]: Will address in my next PR.
